### DISCUSSION TITLE
Add `onOpen`, `onClose` and `onTouch` callbacks

### DISCUSSION THREAD

**Discussion by JesusTheHun** [tailwindlabs/headlessui#1731] [Ideas]
When using a combobox in a form, there is a concept of `touched` fields : if a field is required to not be empty for example, we don't want to show an error message right away when the form loads, it wouldn't make any sense, but once he touches the field (focus the input) and leave the focus with an invalid value (empty input), the field is considered "touched" and therefore, now, we want to show the error message.

Right now we have to use workarounds to make this work, and mark the field as touched : 

`Combobox` : `onBlur` of the `input`
`Listbox` : `onBlur` of the `options`
`Switch` : `onChange` of the `switch`
`RadioGroup` : not possible. `onChange` is only triggered when the value actually changes, so if the user clicks on the selected option, it doesn't work. Using `onClick` on any element below the `RadioGroup` element doesn't work, I assume you stopped the propagation of the `click` event during the capture phase.

Using those hackround to make this work is pretty annoying and requires research on how the `headlessui` components actually work in term of focus management, to get it right. I'm not even sure I got it right yet.

`onTouch` has a strong semantic but is applicable to all components.
`onOpen` & `onClose` are more general purpose callbacks so they could be used for the `touched` state, but they are not applicable in the case of a `Switch` or a `RadioGroup`.

How do you feel about this ?

**Comment by peteruithoven** [comment]
I'd like to add another use case for a Combobox onOpen event. 
The Combobox, when I open it by using the Combobox.Button it puts it in an open mode and clears the ComboBox.Input. If I use the ComboBox.Input to query an server for options based on the input I don't have an event to send a query to get all the options when the input is emptied on open.
I can control the value of ComboBox.Input, but then there is no event to do something with the moment of opening the Combobox.

**Reply by kwojcik** [reply]
We are using Listbox as a multi-select input to filter the contents of a table, whenever the filters are updated we re-request data from the backend. Without an onClose, we have to re-request data on every change instead of when the user is done choosing options which is a perf issue.
